File: server/src/main/java/moe/yuuta/server/topic/TopicExecuteVerticle.java
package moe.yuuta.server.topic;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;

public abstract class TopicExecuteVerticle extends AbstractVerticle {
    static final String EXTRA_TOPIC_ID = TopicExecuteVerticle.class.getName() + ".EXTRA_TOPIC_ID";

    protected String topicId;

    @Override
    public final void start(Future<Void> startFuture) throws Exception {
        topicId = config().getString(EXTRA_TOPIC_ID, null);
        if (topicId == null) {
            startFuture.fail("Topic id is not provided");
            return;
        }
        onRegister(startFuture);
    }

    @Override
    public final void start() throws Exception {
        super.start();
    }

    @Override
    public final void stop() throws Exception {
        super.stop();
    }

    @Override
    public final void stop(Future<Void> stopFuture) throws Exception {
        onUnRegister(stopFuture);
    }

    public void onRegister (Future<Void> registerFuture) throws Exception {
        registerFuture.complete();
    }

    public void onUnRegister (Future<Void> unRegisterFuture) throws Exception {
        unRegisterFuture.complete();
    }
}

File: server/src/main/kotlin/moe/yuuta/server/topic/TopicExecuteVerticle.kt
package moe.yuuta.server.topic

import io.vertx.core.AbstractVerticle
import io.vertx.core.Future

abstract class TopicExecuteVerticle : AbstractVerticle() {
    companion object {
        const val EXTRA_TOPIC_ID = "moe.yuuta.server.topic.TopicExecuteVerticle.EXTRA_TOPIC_ID"
    }

    protected lateinit var topicId: String

    @Throws(Exception::class)
    @Override
    final override fun start(startFuture: Future<Void>) {
        val id: String? = config().getString(EXTRA_TOPIC_ID, null)
        if (id == null) {
            startFuture.fail("Topic id is not provided")
            return
        }
        topicId = id
        onRegister(startFuture)
    }

    @Throws(Exception::class)
    @Override
    final override fun start() {
        super.start()
    }

    @Throws(Exception::class)
    final override fun stop() {
        super.stop()
    }

    @Throws(Exception::class)
    @Override
    final override fun stop(stopFuture: Future<Void>) {
        onUnRegister(stopFuture)
    }

    @Throws(Exception::class)
    open fun onRegister (registerFuture: Future<Void>) {
        registerFuture.complete()
    }

    @Throws(Exception::class)
    open fun onUnRegister (unRegisterFuture: Future<Void>) {
        unRegisterFuture.complete()
    }
}
